### ISSUE TITLE
Example

### ISSUE THREAD

**Issue by joelcnz** [jython234/blocksound#1]
What is an example program? Like loading and playing a sound.

**Comment by alejzeis** [comment]
Sorry, this is my fault for not providing good documentation! But here is an example of playing a basic sound:

``` D
import std.stdio;
import blocksound.core;
import blocksound.audio;

void main() {
    init(); // Make sure to init the library before calling any methods!

    AudioManager mgr = new AudioManager();
    mgr.gain = 1; // Listener Gain
    mgr.listenerLocation = Vec3(0f, 0f, 0f); // Vec3 in blocksound.util (public import in blocksound.core)

    Source source = mgr.createSource(Vec3(0f, 0f, 0f)); // Create a new source at the listener's location
    source.setSound(loadSoundFile("yourfilehere.ogg")); // Set the sound file (loaded into memory UNCOMPRESSED)
    source.play(); // Start playing the sound!

    while(!source.hasFinishedPlaying()) { } // Wait for it to finish

    mgr.deleteSource(source); // Delete the source (frees the sound memory, etc.)
    mgr.cleanup(); // Cleans up the audio manager
}
```

Loading a sound file loads the buffer into memory uncompressed, so if you are playing long audio then be aware of the memory that will be used. I will be adding streaming later, so the sound will be read in chunks from the disk.

File formats supported are listed here (libsndfile): http://www.mega-nerd.com/libsndfile/
Basically it can play WAV, OGG, and FLAC.

If on windows you will need OpenAL and libsndfile binaries (place them in a "lib" directory next to your executable). You can find those [here](http://kcat.strangesoft.net/openal-binaries/openal-soft-1.17.2-bin.zip) and [here](http://www.mega-nerd.com/libsndfile/).

If on linux, make sure to have openal, libsndfile, (and i think libogg, libvorbis) installed.

If on macOS, well, can't really help here. I believe OpenAL is included in system libraries.

**Comment by alejzeis** [comment]
It's a very basic library so far, I plan on adding more features in time.

**Comment by alejzeis** [comment]
@joelcnz The version on DUB (1.0.0) probably won't work with my example. I've created a new release that will: 1.0.0-alpha2, but it will take a while for DUB to update it's listings. In the meantime you can use `"~master"` as the version in your `dub.json`.

**Comment by joelcnz** [comment]
O, I'm using Mac, and get an errors about OpenAL.

**Comment by alejzeis** [comment]
@joelcnz Could you please post the errors? It shouldnt be too hard to fix.

**Comment by joelcnz** [comment]
Here is what I get:

Joels-MacBook-Pro:BlockSound joelcnz$ dub
WARNING: A deprecated branch based version specification is used for the dependency blocksound. Please use numbered versions instead. Also note that you can still use the dub.selections.json file to override a certain dependency to use a branch instead.
Performing "debug" build using dmd for x86_64.
derelict-util 2.0.6: target for configuration "library" is up to date.
derelict-al 1.0.1: target for configuration "library" is up to date.
derelict_extras-sndfile 2.0.1: target for configuration "library" is up to date.
blocksound ~master: target for configuration "openal-backend" is up to date.
soundblock ~master: target for configuration "application" is up to date.
To force a rebuild of up-to-date targets, run again with --force.
Running ./soundblock 
derelict.util.exception.SharedLibLoadException@../../../../.dub/packages/derelict-util-2.0.6/source/derelict/util/exception.d(35): Failed to load one or more shared libraries:
    libsndfile.1.dylib - dlopen(libsndfile.1.dylib, 2): image not found

##     libsndfile.dylib - dlopen(libsndfile.dylib, 2): image not found

4   soundblock                          0x00000001076528a1 void derelict.util.exception.SharedLibLoadException.throwNew(immutable(char)[][], immutable(char)[][]) + 433
5   soundblock                          0x000000010765639c void derelict.util.sharedlib.SharedLib.load(immutable(char)[][]) + 484
6   soundblock                          0x0000000107652e49 void derelict.util.loader.SharedLibLoader.load(immutable(char)[][]) + 117
7   soundblock                          0x0000000107652d41 void derelict.util.loader.SharedLibLoader.load(immutable(char)[]) + 301
8   soundblock                          0x0000000107652b99 void derelict.util.loader.SharedLibLoader.load() + 101
9   soundblock                          0x0000000107645066 void blocksound.backend.openal.loadLibraries(bool, bool) + 62
10  soundblock                          0x0000000107645a99 @trusted void blocksound.core.init() + 13
11  soundblock                          0x0000000107643b4c _Dmain + 12
12  soundblock                          0x0000000107673bc3 D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv + 39
13  soundblock                          0x0000000107673af7 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 35
14  soundblock                          0x0000000107673b68 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])_).runAll() + 44
15  soundblock                          0x0000000107673af7 void rt.dmain2._d_run_main(int, char__, extern (C) int function(char[][])_).tryExec(scope void delegate()) + 35
16  soundblock                          0x0000000107673a5d _d_run_main + 497
17  soundblock                          0x0000000107643d29 main + 33
18  libdyld.dylib                       0x00007fff9cc955ac start + 0
19  ???                                 0x0000000000000000 0x0 + 0
Program exited with code 1
Joels-MacBook-Pro:BlockSound joelcnz$ 

Thanks.

**Comment by alejzeis** [comment]
@joelcnz That should be an easy fix, BlockSound can't find the libsndfile dylib file. I do not have a macOS system avaliable so I will try to guide you on how to obtain it.

First you will need [homebrew](http://brew.sh/), if you don't have it. You can run the terminal command from the webpage to install it.

Once it is installed you can run `brew install libsndfile` and it should work after that.

**Comment by alejzeis** [comment]
If you want to distribute your application and don't want users to have to go through the homebrew process, you should be able to copy the libsndfile .dylib file from homebrew's library directory and distribute that with your executable. OpenAL is included by default on most macOS systems.

**Comment by joelcnz** [comment]
Thanks.

I think my home brew is broken. I'm meaning to reinstall my OSX. I've got a SSD ready to go, but I'm  waiting for my brother to help me install it and stuff.

**Comment by alejzeis** [comment]
@joelcnz I see. Well, if you need any more help, just let me know.
